tokenizer_ts/src/bytePairEncode.ts

```diff
@@ -20,66 +73,67 @@ export function uint8ArrayToString(uint8Array: Uint8Array): string {
  */
 export function bytePairEncode(
   mergingBytes: Uint8Array,
-  ranks: ReadonlyMap<string, number>
+  ranks: BinaryMap<number>
 ): number[] {
   if (mergingBytes.length === 1) {
-    return [ranks.get(mergingBytes[0].toString())!];
+    return [ranks.get(mergingBytes)!];
   }
 
+  let minRank = maxRank;
+  let minIndex = -1;
+
   const byteIndicesAndRanks: [number, number][] = [];
-  for (let i = 0; i < mergingBytes.length + 1; i++) {
-    byteIndicesAndRanks.push([i, Number.MAX_SAFE_INTEGER]);
+  for (let i = 0; i < mergingBytes.length - 1; i++) {
+    const rank = ranks.get(mergingBytes.subarray(i, i + 2)) ?? maxRank;
+    if (rank < minRank) {
+      minRank = rank;
+      minIndex = i;
+    }
+
+    byteIndicesAndRanks.push([i, rank]);
   }
+  byteIndicesAndRanks.push([mergingBytes.length - 1, maxRank]);
+  byteIndicesAndRanks.push([mergingBytes.length, maxRank]);
 
   function getRank(startIndex: number, skip = 0): number {
     if (startIndex + skip + 2 < byteIndicesAndRanks.length) {
-      const slice = mergingBytes.slice(
+      const slice = mergingBytes.subarray(
```

Comment:
Since this is a very hot section of the code, instead of creating a subarray can we pass `start, end` offset arguments to `get` and use those? Even though `subarray` doesn't copy data it still creates a new object. Microbenchmark shows using `subarray` is a good deal slower if we're just reading a slice of the array

Reply:
Good idea! I'll do that in a followup (unless you or @andreamah wants to take it!)

Reply:
Sure, I can try to take it! I'm not sure what you mean by `get(start,end)`, though?

Reply:
I think he means to make `ranks.get(key: Uint8Array)`, to `ranks.get(key: Uint8Array, start: number, end: number)` so that we don't need to call `.subarray()` at all

Reply:
Just so I understand correctly, is the point of passing `start, end` into the `BinaryMap` such that we can just start looking at `start` index and then only traverse for `end-start` elements?

Reply:
Also, on top of this: is there an efficient way to start at `start` when you're traversing through the multi-level map? For example, I'm not sure if Uint8Array has something where we can bit shift the first `n` numbers out from the start of the array, hence creating a new start and length?

Reply:
I think you would also change `binaryMapKey`'s signature and instead of getting `0`, `1`, `2`, etc. you get `start + 0`, `start + 1`, ...

Reply:
Btw, currently `binaryMapKey` always looks at 6 bytes, depending on the out-of-bounds behavior of typed arrays to return 0. But you can no longer depend on this if we just give it a start and end. However, we can fix this and avoid adding any conditional branches (which run slower in spite of [branch prediction](https://blog.cloudflare.com/branch-predictor)) by creating binary masks that remove any bytes from outside our range:

```ts
const binaryMapKey = (k: Uint8Array, start: number, end: number): number => {
  const length = end - start;
  
  // 'lower' and 'upper' are both 24-bit integers, like
  //    0xFF FF FF
  //      ^3 ^2 ^1
  // If we say have a length of 2, we should disregard the last "3" byte, so we
  // create a mask like
  //    0x00 FF FF  (started at 0xFF FF FF and shifted over by 8 bits)
  //      ^3 ^2 ^1
  // so that we discard the data outside our range
  const lowerMask = 0xFFFFFF >>> Math.max(0, (3 - length) * 8);
  const lower = (k[start + 0] | (k[start + 1] << 8) | (k[start + 2] << 16)) & lowerMask;
  const upperMask = 0xFFFFFF >>> Math.max(0, (6 - length) * 8);
  const upper = (k[start + 3] | (k[start + 4] << 8) | (k[start + 5] << 16)) & upperMask;
  return lower + (0xFFFFFF * upper);
};
```

(have not tested this, but something along these lines should work)

Reply:
Oh, that's so smart! I was thinking of a more branch-based approach with ternary operators, but this seems way faster

Reply:
I incorporated your code into the BinaryMap and made the changes that I thought were needed for supporting a start/end, but I seem to still have a bug somewhere https://github.com/andreamah/Tokenizer/pull/4. I can take a look again later, but am I misunderstanding something about the map?

Reply:
That looks right. It could be that I have a bug in my proposed function or something. Perhaps it's worth adding a few standalone tests for the BinaryMap to help diagnose that.